Index: torch-ccl/third_party/oneCCL/src/CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- torch-ccl/third_party/oneCCL/src/CMakeLists.txt	(revision 15dafd4094016c50a571b4e891aa344db090e7f2)
+++ torch-ccl/third_party/oneCCL/src/CMakeLists.txt	(date 1608299636844)
@@ -232,7 +232,7 @@
 add_library(ccl SHARED $<TARGET_OBJECTS:ccl-objects>)
 target_include_directories(ccl PUBLIC ${CCL_INC_DIRS})
 # link with release_mt libmpi.so for oneAPI Base toolkit
-set_target_properties(ccl PROPERTIES LINK_FLAGS "-Wl,-rpath,../../../../mpi/latest/lib/release_mt/")
+set_target_properties(ccl PROPERTIES LINK_FLAGS "-Wl,-rpath,${CMAKE_INSTALL_RPATH}")
 target_link_libraries(ccl PUBLIC dl pthread ${EXTERNAL_LIBS} ${COMPUTE_RUNTIME_TARGET_NAME} fabric mpi)
 if (NOT LIB_SO_VERSION AND NOT LIB_MAJOR_VERSION)
         set_target_properties(ccl PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CCL_BUILD_DIR})
@@ -240,6 +240,7 @@
         set_target_properties(ccl PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CCL_BUILD_DIR} VERSION ${LIB_SO_VERSION} SOVERSION ${LIB_MAJOR_VERSION})
 endif()
 install(TARGETS ccl LIBRARY DESTINATION ${CCL_INSTALL_LIB})
+
 install(FILES
     "../cmake/FindOpenCL.cmake"
     "../cmake/FindL0.cmake"
