diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index abe7d962..d5474ed2 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -276,6 +276,8 @@ endif()
 # shared library
 add_library(ccl SHARED $<TARGET_OBJECTS:ccl-objects>)
 target_include_directories(ccl PUBLIC ${SRC_INCLUDE_DIRS})
+set(ONEAPI_IMPI_RPATH "'$ORIGIN'")
+set_target_properties(ccl PROPERTIES LINK_FLAGS "-Wl,-rpath,${ONEAPI_IMPI_RPATH}")
 target_link_libraries(ccl PRIVATE ${SRC_LINK_LIBS})
 
 set_target_properties(ccl PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CCL_BUILD_DIR})
