diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 91d13d05..320be163 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -279,7 +279,7 @@ endif()
 add_library(ccl SHARED $<TARGET_OBJECTS:ccl-objects>)
 target_include_directories(ccl PUBLIC ${SRC_INCLUDE_DIRS})
 
-target_link_libraries(ccl PUBLIC ${SRC_LINK_LIBS})
+target_link_libraries(ccl PRIVATE ${SRC_LINK_LIBS})
 
 if (NOT LIB_SO_VERSION AND NOT LIB_MAJOR_VERSION)
     set_target_properties(ccl PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CCL_BUILD_DIR})
